src/misc/setup.cpp

```diff
@@ -1248,56 +1250,84 @@ bool Config::ParseConfigFile(const std::string& type, const std::string& configf
 		return false;
 	}
 
-	configfiles.push_back(configfilename);
+	configfiles.push_back(config_file_name);
 	configFilesCanonical.push_back(canonical_path);
 
-	// Get directory from configfilename, used with relative paths.
+	// Get directory from config_file_name, used with relative paths.
 	current_config_dir = canonical_path.parent_path().string();
 
-	string line;
-	Section* currentsection = nullptr;
+	// If this is an autoexec section, the above takes care of the joining
+	// while this handles the overwrriten mode. We need to be prepared for
+	// either scenario to play out because we won't know the users final
+	// preference until the very last configuration file is processed.
+
+	std::string line = {};
+
+	Section* current_section = nullptr;
+	bool is_autoexec_section = false;
+	bool is_autoexec_started = false;
+
+	auto is_empty_line = [](const std::string& line) {
+		return line.empty() || line[0] == '\0' || line[0] == '\n' ||
+		       line[0] == '\r';
+	};
+
+	auto is_comment = [](const std::string& line) {
+		return !line.empty() && (line[0] == '%' || line[0] == '#');
+	};
+
+	auto is_section_start = [](const std::string& line) {
+		return !line.empty() && line[0] == '[';
+	};
+
+	auto handle_autoexec_line = [&]() {
+		// Ignore all the empty lines until the meaningful [autoexec]
+		// content starts
+		if (!is_autoexec_started) {
+			if (is_empty_line(line) || is_comment(line)) {
+				return;
+			}
+			is_autoexec_started = true;
+		}
+
+		if (!is_comment(line)) {
+			current_section->HandleInputline(line);
+			OverwriteAutoexec(config_file_name, line);
+		}
+	};
 
 	while (getline(in, line)) {
-		// Strip leading/trailing whitespace
 		trim(line);
-		if (line.empty()) {
+
+		// Special handling of [autoexec] section
+		if (is_section_start(line)) {
+			is_autoexec_section = false;
+			is_autoexec_started = false;
+		}
+		if (is_autoexec_section) {
+			handle_autoexec_line();
 			continue;
 		}
```

Comment:
Maybe move the special handling to the comment, and because it's a separate branch from the `is_section_start` flow, it can be first too:

```c++
		// Special handling of [autoexec] section
		if (is_autoexec_section) {
			handle_autoexec_line();
			continue;
		}

		if (is_section_start(line)) {
			is_autoexec_section = false;
			is_autoexec_started = false;
		}
```

Reply:
Comment moved. Reordering the conditions is not possible - if someone starts new section after `[autoexec]`, we need to stop inserting lines into autoexec immediately.